File: mod/ValeLoot/LootFilter.cs
using System;

namespace ValeLoot;

/// <summary>
/// The loot filter model, and the evaluation that turns an item into a colour.
///
/// This is a port of the overlay's `loot-filter.ts` semantics into the mod, and the port exists for
/// one reason: THE MOD MUST BE COMPLETE ON ITS OWN. A player who has never heard of the overlay
/// installs the plugin, edits a text file, and their bag lights up. Nothing is pushed in over a
/// socket, nothing else has to be running, and the mod holds the rules and does the judging.
///
/// The semantics are deliberately identical to the overlay's, so one filter file reads the same in
/// both places: an ORDERED list of rules, FIRST MATCH WINS, each rule deciding how an item is
/// PRESENTED — a colour, a short tag, how hard the cell is highlighted, and whether it makes a noise
/// when it arrives. Ordering is the whole ergonomic trick: put "triple top roll" above "vendor trash"
/// and stop thinking about overlap.
///
/// ## Presentation is the entire vocabulary, deliberately
///
/// A rule cannot ACT on an item. There is no field here that could express any automation, and there
/// is no code in this plugin that could carry one out. That is a product decision, not an omission:
/// the game's staff allow client-side tools that change "how the loot color is, sound and how it
/// looks in the inventory" and nothing past that line. Keeping the TYPE incapable of an action is
/// cheaper than keeping a policy — nothing downstream has to be trusted to check.
///
/// ## Rolls, and the two questions the `%` separates
///
/// `StatData.Value` IS the roll percentage, 0..100 — not the number the game prints. The game derives
/// what it prints: `displayed = cap * (2/3 + roll/300)`, rounded. So "top roll" questions are free and
/// in-process, and `TopRolls`, `AvgRoll` and `Stat Agi >= 90%` all work from the item alone. The
/// ABSOLUTE form, `Stat Agi >= 3`, needs the item's base cap, which <see cref="ItemCatalog"/> reads
/// out of the game's own configs.
///
/// The two stay SEPARATE fields on <see cref="StatCondition"/> rather than one number with a flag,
/// because they are different questions and the whole point of the `%` suffix is that the player
/// chose which one to ask. A filter that quietly answers the other one grades every item wrong.
///
/// ## Cost
///
/// Evaluation runs over every visible cell on every inventory repaint, inside a UI callback that must
/// not stutter. So it allocates nothing per item: no lower-casing (comparisons are
/// OrdinalIgnoreCase), no substring building, no regex. <see cref="ItemFacts"/> is a single buffer the
/// reader refills per cell rather than an object per item.
/// </summary>
internal static class LootFilter
{
    /// <summary>Most substat lines one item can carry. Bounds the fact buffer; overflow is dropped.</summary>
    public const int MaxStats = 16;

    /// <summary>Roll percentage that counts as a top roll unless the filter says otherwise.</summary>
    public const int DefaultThreshold = 90;

    /**
     * One item, flattened to what a rule can ask about — refilled per cell, never allocated per cell.
     *
     * `StatRolls` holds `StatData.Value`, which is the ROLL PERCENTAGE. It can exceed 100: the game
     * branches on `Value > 100` for the chaos over-roll case, so the array is not clamped and
     * <see cref="LootCondition.OverRoll"/> exists to ask about exactly that.
     *
     * `StatTiers` is the quality-tier string the game itself prints for the line. Nothing filters on it
     * — inventing a condition on a string whose vocabulary we have not read would be a guess — but it
     * is what makes a `probe` reply describe a real item rather than a row of numbers.
     */
    internal sealed class ItemFacts
    {
        /// <summary>The name the player sees on the cell.</summary>
        public string Name = "";
        /// <summary>`InventoryItemData.Id` — the catalog id, which a filter may also match on.</summary>
        public string Id = "";
        /// <summary>The type the cell displays ("Accessory", "Rifle", …) — what the player reads.</summary>
        public string Type = "";
        public int Refine;
        public bool Favorite;
        public bool HasChaos;

        public int StatCount;
        public readonly string[] StatNames = new string[MaxStats];
        /// <summary>`StatType` ordinals, parallel to the names — what the catalog keys a cap by.</summary>
        public readonly int[] StatTypes = new int[MaxStats];
        public readonly int[] StatRolls = new int[MaxStats];
        public readonly string[] StatTiers = new string[MaxStats];

        public void Reset()
        {
            Name = "";
            Id = "";
            Type = "";
            Refine = 0;
            Favorite = false;
            HasChaos = false;
            StatCount = 0;
        }

        public void AddStat(string name, int statType, int roll, string tier)
        {
            if (StatCount >= MaxStats) return;
            StatNames[StatCount] = name;
            StatTypes[StatCount] = statType;
            StatRolls[StatCount] = roll;
            StatTiers[StatCount] = tier;
            StatCount++;
        }

        /// <summary>Lines rolling at or above the threshold — the "triple top roll" count.</summary>
        public int TopRolls(int threshold)
        {
            int count = 0;
            for (int i = 0; i < StatCount; i++) if (StatRolls[i] >= threshold) count++;
            return count;
        }

        /**
         * Mean roll, rounded to a whole percent, or -1 when the item has no lines.
         *
         * Rounded, and then compared as a whole number, so `AvgRoll < 35` means exactly what a player
         * gets by averaging the percentages they can see. The overlay compares unrounded and nudges its
         * strict bounds by 1e-9 because it draws the unrounded figure in its own HUD; in here there is
         * no such HUD, so a boundary the player cannot reproduce by hand would be a boundary they
         * cannot trust.
         *
         * No lines is -1 rather than 0: "average roll below 35" is a statement about an item that rolled
         * badly, not about one that cannot roll at all.
         */
        public int AverageRoll()
        {
            if (StatCount == 0) return -1;
            int sum = 0;
            for (int i = 0; i < StatCount; i++) sum += StatRolls[i];
            return (int)Math.Round(sum / (double)StatCount, MidpointRounding.AwayFromZero);
        }

        public bool HasOverRoll()
        {
            for (int i = 0; i < StatCount; i++) if (StatRolls[i] > 100) return true;
            return false;
        }
    }

    /**
     * One required substat line, in one of its two forms: `Stat Agi >= 90%` or `Stat Agi >= 3`.
     *
     * The `%` form is a floor on ROLL QUALITY, which the item carries. The bare form is a floor on
     * the VALUE THE GAME PRINTS, which needs the item's base cap out of the catalog. They are two
     * fields rather than one number and a flag on purpose: they are different questions, the parser
     * sets exactly one of them per line, and nothing downstream can mix them up by accident.
     */
    internal sealed class StatCondition
    {
        public string Stat = "";
        /// <summary>Roll-quality floor in percent — the `%` form. Null when the line did not ask.</summary>
        public int? MinRollPct;
        /// <summary>Printed-value floor — the bare form. Null when the line did not ask.</summary>
        public int? MinValue;
    }

    /// <summary>What a rule tests. Every field absent means "any item", which only a Show block may mean.</summary>
    internal sealed class LootCondition
    {
        public string[]? Types;
        /**
         * Any of these, as a case-insensitive substring of the name, the id or the catalog's name.
         *
         * A LIST rather than one string because the commonest real rule is "these four drops I care
         * about", and the alternative is four `Show` blocks that differ only in one word — four
         * places to change the colour, and four chances to get the order wrong. `Type` already reads
         * as a list for the same reason, so `Name "Buzzing Hive Fragment", "Abyssal Idol"` is the
         * spelling a player would guess.
         */
        public string[]? Names;
        public int? MinRefine;
        public int? MinTopRolls;
        public int? MaxTopRolls;
        public int? MinAvgRoll;
        public int? MaxAvgRoll;
        public StatCondition[]? Stats;
        /// <summary>Every listed stat must match, rather than just one. `AnyStat` flips it.</summary>
        public bool StatsAll = true;
        public bool? HasChaos;
        public bool? Favorite;
        /// <summary>A line rolled past 100 — the chaos over-roll, and unambiguously a great item.</summary>
        public bool? OverRoll;

        public bool IsEmpty =>
            Types is null && Names is null && MinRefine is null
            && MinTopRolls is null && MaxTopRolls is null && MinAvgRoll is null && MaxAvgRoll is null
            && (Stats is null || Stats.Length == 0)
            && HasChaos is null && Favorite is null && OverRoll is null;
    }

    /**
     * How hard a match shouts in the bag, as the level the cell overlay is driven to.
     *
     *   dot (1)   quiet — the cell carries the rule's colour, nothing more
     *   mark (2)  clearly lit: worth a second look
     *   glow (3)  unmistakable
     *
     * Three levels rather than a boolean, because a bag where 25 of 31 slots shout says nothing.
     */
    public const int LevelDot = 1;
    public const int LevelMark = 2;
    public const int LevelGlow = 3;

    /// <summary>`dot`/`mark`/`glow` -> level, or 0 for anything else.</summary>
    public static int ParseLevel(string word) => word switch
    {
        "dot" => LevelDot,
        "mark" => LevelMark,
        "glow" => LevelGlow,
        _ => 0,
    };

    public static string LevelName(int level) => level switch
    {
        LevelGlow => "glow",
        LevelMark => "mark",
        LevelDot => "dot",
        _ => "none",
    };

    /**
     * `#rrggbb` -> 0..1 floats, falling back to white.
     *
     * No gamma conversion: the colour is the one written in the filter file, and a "more correct"
     * transform here would make the cell disagree with the hex the player typed.
     *
     * Parsed once per rule at load rather than per cell per repaint — see `LootRule.R/G/B`.
     */
    public static (float R, float G, float B) ParseColor(string? hex)
    {
        if (hex is null) return (1f, 1f, 1f);
        int start = hex.Length > 0 && hex[0] == '#' ? 1 : 0;
        if (hex.Length - start < 6) return (1f, 1f, 1f);
        try
        {
            int value = Convert.ToInt32(hex.Substring(start, 6), 16);
            return (((value >> 16) & 0xFF) / 255f, ((value >> 8) & 0xFF) / 255f, (value & 0xFF) / 255f);
        }
        catch { return (1f, 1f, 1f); }
    }

    internal sealed class LootRule
    {
        public string Name = "";
        public string Color = "#4ade80";
        /// <summary>`Color`, pre-parsed at load. A repaint must not parse a hex string per cell.</summary>
        public float R = 0.29f;
        public float G = 0.87f;
        public float B = 0.5f;
        public string Label = "";
        public int Level = LevelDot;
        /// <summary>Played once when a match first appears in the bag. Null for silence.</summary>
        public string? Sound;
        /**
         * Matched, and deliberately silent — the filter language's `Hide`.
         *
         * A muted match draws nothing and plays nothing. It is NOT "no rule matched": the rule claimed
         * the item and chose silence, which is how a filter stops lighting up vendor fodder while still
         * letting a later rule be quiet about nothing.
         */
        public bool Mute;
        public LootCondition When = new();
        /// <summary>Source line of the block opener, so a message can point at the player's own file.</summary>
        public int Line;
    }

    /// <summary>
    /// First matching rule, or null. Order is the filter's meaning, so this never reorders or scores.
    ///
    /// `AlwaysShow`/`AlwaysHide` are applied by the caller BEFORE this, because a per-item override
    /// exists precisely to escape rule order.
    /// </summary>
    public static LootRule? Match(ItemFacts item, LootRule[] rules, int threshold)
    {
        for (int i = 0; i < rules.Length; i++)
        {
            if (Matches(item, rules[i].When, threshold)) return rules[i];
        }
        return null;
    }

    public static bool Matches(ItemFacts item, LootCondition when, int threshold)
    {
        if (when.Types is not null)
        {
            // The cell's text OR the catalog's own `EquipType` member name. Both, rather than the
            // catalog winning, because a filter written against what the cell prints must keep
            // working — and the enum name is what covers a cell whose text is localised.
            string catalogType = ItemCatalog.TypeName(item.Id) ?? "";
            bool hit = false;
            for (int i = 0; i < when.Types.Length; i++)
            {
                string want = when.Types[i];
                if (string.Equals(item.Type, want, StringComparison.OrdinalIgnoreCase)
                    || (catalogType.Length > 0 && string.Equals(catalogType, want, StringComparison.OrdinalIgnoreCase)))
                {
                    hit = true;
                    break;
                }
            }
            if (!hit) return false;
        }

        // Matched against the DISPLAYED name, the catalog id, and the config's own display name.
        // Players think in display names ("Kunai"), filters shared from elsewhere may carry ids, and
        // asking which one a given string is would make the commonest line in any filter the one that
        // needs explaining. The config name is the third because a cell truncates its text to fit, a
        // pickup has no cell at all, and several kinds of item have an id nothing like their name —
        // "Buzzing Hive Fragment" is `Lure Sting`.
        //
        // ANY of the listed names is enough. Every one of them asks the same three questions, so a
        // list is exactly as strong as the `Show` blocks it replaces.
        if (when.Names is not null)
        {
            string catalogName = ItemCatalog.DisplayName(item.Id) ?? "";
            bool named = false;
            for (int i = 0; i < when.Names.Length; i++)
            {
                string want = when.Names[i];
                if (item.Name.IndexOf(want, StringComparison.OrdinalIgnoreCase) >= 0
                    || item.Id.IndexOf(want, StringComparison.OrdinalIgnoreCase) >= 0
                    || (catalogName.Length > 0
                        && catalogName.IndexOf(want, StringComparison.OrdinalIgnoreCase) >= 0))
                {
                    named = true;
                    break;
                }
            }
            if (!named) return false;
        }

        if (when.MinRefine is int minRefine && item.Refine < minRefine) return false;
        if (when.HasChaos is bool chaos && item.HasChaos != chaos) return false;
        if (when.Favorite is bool favorite && item.Favorite != favorite) return false;
        if (when.OverRoll is bool over && item.HasOverRoll() != over) return false;

        if (when.MinTopRolls is not null || when.MaxTopRolls is not null)
        {
            int top = item.TopRolls(threshold);
            if (when.MinTopRolls is int minTop && top < minTop) return false;
            if (when.MaxTopRolls is int maxTop && top > maxTop) return false;
        }

        if (when.MinAvgRoll is not null || when.MaxAvgRoll is not null)
        {
            int average = item.AverageRoll();
            if (average < 0) return false;
            if (when.MinAvgRoll is int minAvg && average < minAvg) return false;
            if (when.MaxAvgRoll is int maxAvg && average > maxAvg) return false;
        }

        if (when.Stats is not null && when.Stats.Length > 0)
        {
            int hits = 0;
            for (int i = 0; i < when.Stats.Length; i++)
            {
                if (MatchesStat(item, when.Stats[i])) hits++;
                else if (when.StatsAll) return false;
            }
            if (!when.StatsAll && hits == 0) return false;
        }

        return true;
    }

    private static bool MatchesStat(ItemFacts item, StatCondition want)
    {
        for (int i = 0; i < item.StatCount; i++)
        {
            if (!string.Equals(item.StatNames[i], want.Stat, StringComparison.OrdinalIgnoreCase)) continue;
            if (want.MinRollPct is int minRoll && item.StatRolls[i] < minRoll) continue;
            if (want.MinValue is int minValue)
            {
                // An unanswerable condition is NOT a match. `TryScaledValue` returns false when the
                // catalog has not resolved, when the item is not one it knows, or when the stat is
                // not in that item's pool — and treating any of those as satisfied would widen the
                // block to the whole bag. It says so in the log once, and declines here.
                if (!ItemCatalog.TryScaledValue(item.Id, item.StatTypes[i], item.StatRolls[i], out int printed)) continue;
                if (printed < minValue) continue;
            }
            return true;
        }
        return false;
    }
}

File: src/filter/loot-filter.ts
/*
 * VENDORED. Copied from the private project this filter language was written for, where the
 * canonical copy lives beside the rest of that project's bridge modules. It is here because the rule
 * editor in `tools/valeloot-editor/` compiles the REAL matcher into its page rather than a lookalike:
 * a second implementation of these semantics would make the editor agree with the mod only by
 * coincidence. Both copies are MIT and ours.
 *
 * Changed from the original: the sibling `import type`s now come from `./types.ts` (see that file),
 * and one citation of an unpublished design document was dropped. Nothing executable was touched —
 * fix bugs in the canonical copy and re-copy, do not diverge here.
 */
/**
 * Loot filters — your rules, your colours, your sounds.
 *
 * Modelled on the loot filters players already know (Path of Exile, Diablo): an ORDERED list of rules,
 * FIRST MATCH WINS, each rule deciding how an item is PRESENTED — a colour, a short tag, how hard it is
 * highlighted in the bag, and whether it makes a noise when it lands. Rules can also `mute`: PoE's
 * `Hide`, for the trash you want the overlay to stop talking about.
 *
 * That ordering is the whole ergonomic trick — you put "triple top roll" above "vendor trash" and stop
 * thinking about overlap.
 *
 * ## Presentation is the entire vocabulary, deliberately
 *
 * A rule cannot ACT on an item. There is no field here that could express dismantling, pickup or any
 * other automation, and the destructive planner that used to sit next to this file is gone. That is a
 * product decision, not an omission: the game's staff allow client-side tools that change "how the loot
 * color is, sound and how it looks in the inventory" and nothing past that line. Keeping the *type*
 * incapable of an action is cheaper than keeping a policy: nothing downstream has to be trusted to
 * check.
 *
 * What a rule can test is deliberately limited to things that are TRUE about the item: its type, its
 * refine, its substat lines and their roll percentages, how many lines clear your top-roll threshold,
 * whether the wiki knows it, and how it compares to what you are wearing. There is no "is it good"
 * heuristic hidden in here — good is whatever your rules say it is.
 *
 * Evaluation is pure and allocation-light: it runs over a few hundred items every time a bag changes,
 * inside an overlay that must not stutter, so it does no string building and no regex compilation per
 * item.
 */

import type { OwnedGear, Verdict } from './types.ts';

export interface StatCondition {
  stat: string;
  /** The line must roll at least this well (0..100). Omit to accept any roll. */
  minRollPct?: number;
  /**
   * The line's VALUE as the game prints it must be at least this — the `3` in "+3 AGI".
   *
   * Distinct from `minRollPct`, which is where the roll sits in the stat's legal range. A player
   * asking for "kunais with +3 AGI" means this one; "top-roll AGI" means the other. Conflating them
   * silently answers a different question than the one asked.
   */
  minValue?: number;
}

export interface LootCondition {
  /** Item types this rule applies to (`Chest`, `Pistol`, …). Empty/absent = any. */
  slotTypes?: string[];
  /** At least this many lines at or above the inventory threshold. */
  minTopRolls?: number;
  /** At most this many. `maxTopRolls: 0` is "nothing good on it" — the core of any trash rule. */
  maxTopRolls?: number;
  minAvgRoll?: number;
  maxAvgRoll?: number;
  minRefine?: number;
  /** Required substat lines. */
  stats?: StatCondition[];
  /** Whether every listed stat must be present, or just one. Default 'all'. */
  statMode?: 'all' | 'any';
  /** At least this many of the item's stats are ones your worn gear already uses. */
  minSharedStats?: number;
  /** Only items the site's catalog does not know (or only ones it does). */
  unknown?: boolean;
  /** Case-insensitive substring of the item name. */
  nameContains?: string;
  /** Only items whose upgrade comparison reached one of these verdicts. */
  verdicts?: Verdict[];
  /** Only items with a chaos substat (or only ones without). */
  hasChaos?: boolean;
  /**
   * Only items the player has flagged as a favourite in game (or only ones they have not).
   *
   * The game's own gesture for "don't touch this", and the cheapest signal in the bag because the
   * player already maintains it. Present here because the in-game mod reads the same flag off the item
   * and its starter filter uses it — a condition one side has and the other does not makes a shared
   * file unusable in one of them.
   */
  favorite?: boolean;
  /**
   * Only items with a line above its normal maximum (or only ones without).
   *
   * Reachable only by a Chaos widen, so it is a small and unambiguously interesting population — the
   * client itself branches on the roll exceeding 100. Distinct from `hasChaos`, which says the item HAS
   * a chaos slot; this says the chaos actually bought something.
   */
  overRoll?: boolean;
}

/**
 * How hard a match shouts in the bag.
 *
 *   dot    quiet — the cell carries the rule's colour and tag, nothing more
 *   mark   earns a keep mark: the star the grid draws over slots worth a second look
 *   glow   an animated ring, and the loot row pulses as the item lands
 *
 * Three levels rather than a boolean because the previous boolean (`flash`) had to mean both "worth a
 * star" and "worth an animation", and a bag where 25 of 31 slots are starred says nothing.
 */
export type LootHighlight = 'dot' | 'mark' | 'glow';
export const LOOT_HIGHLIGHTS: readonly LootHighlight[] = ['dot', 'mark', 'glow'];

/**
 * Sounds the overlay can make with nothing on disk: each is synthesised in the HUD's audio graph, so a
 * fresh install has usable sounds and this repo ships no binary assets.
 *
 * Any other value in `LootRule.sound` names a file the player dropped in `<settings dir>/sounds` —
 * "custom sounds" means their file, not a curated pack we have to license.
 */
export const BUILTIN_LOOT_SOUNDS: readonly string[] = ['blip', 'chime', 'ding', 'alert', 'thud'];

/**
 * A sound name that is safe to put in a URL and to resolve inside the sounds directory, or null.
 *
 * Rules are hand-editable JSON and pasted filter text, so this is the one place a name is vetted:
 * no separators, no `..`, no spaces. The HTTP layer resolves files by this name, and a filter that
 * could name `../../secrets` would make the editor a file reader.
 */
export function normalizeSoundName(input: unknown): string | null {
  if (typeof input !== 'string') return null;
  const value = input.trim();
  return /^[A-Za-z0-9][A-Za-z0-9._-]{0,39}$/.test(value) ? value : null;
}

export interface LootRule {
  id: string;
  name: string;
  enabled: boolean;
  /** Hex colour used for the border, tag and text in the overlay. */
  color: string;
  /** Short tag, e.g. "KEEP". Falls back to the rule name. */
  label?: string;
  /** How loudly to draw a match. Absent means `dot`. */
  highlight?: LootHighlight;
  /**
   * Play this once when a match ARRIVES in the bag — a built-in name or a file in the sounds directory.
   *
   * On arrival only, never on a rescan: the bag is re-evaluated on every exp tick, and a sound that
   * fired per evaluation would be a siren. `Session` emits loot events from a bag DIFF, which is what
   * makes "on arrival" a fact rather than a debounce.
   */
  sound?: string;
  /**
   * Matched, and deliberately silent — PoE's `Hide`.
   *
   * A muted match draws no mark, plays no sound and generates no loot row. It is not "no rule matched":
   * the rule claimed the item and chose silence, which is how a filter stops the overlay narrating
   * every piece of vendor fodder while still letting a later rule be quiet about nothing.
   */
  mute?: boolean;
  when: LootCondition;
}

/** Everything a rule may need to know that is not on the item itself. */
export interface LootContext {
  /** Roll percentage that counts as a top roll — the same threshold the inventory view used. */
  threshold: number;
  /** Stats your worn gear uses, for `minSharedStats`. */
  wornStats?: ReadonlySet<string>;
  /** Upgrade verdicts by item uid, for `verdicts`. */
  verdictByUid?: ReadonlyMap<string, Verdict>;
}

export interface LootMatch {
  ruleId: string;
  name: string;
  color: string;
  label: string;
  highlight: LootHighlight;
  /** Null when the rule asks for no sound — the common case. */
  sound: string | null;
  mute: boolean;
}

/**
 * Sensible defaults, ordered the way a player would order them. Everything here is editable and
 * removable — these exist so the overlay is useful the first time it runs, not to encode taste.
 *
 * Only the two rules a player would want to hear about carry a sound, and nothing is muted: a first
 * run should show what it can see, and going quiet is a choice the player makes about their own bag.
 */
export function defaultLootRules(): LootRule[] {
  return [
    {
      id: 'triple', name: 'Triple top roll', enabled: true, color: '#e9c46a', label: 'TRIPLE',
      highlight: 'glow', sound: 'chime',
      when: { minTopRolls: 3 },
    },
    {
      id: 'upgrade', name: 'Beats what I wear', enabled: true, color: '#4ade80', label: 'UPGRADE',
      highlight: 'glow', sound: 'ding',
      when: { verdicts: ['upgrade'] },
    },
    {
      id: 'double', name: 'Two top rolls', enabled: true, color: '#7cc0ff', label: 'KEEP',
      highlight: 'mark',
      when: { minTopRolls: 2 },
    },
    {
      id: 'mystat', name: 'Rolls my stats well', enabled: true, color: '#a78bfa', label: 'MINE',
      highlight: 'mark',
      when: { minSharedStats: 2, minAvgRoll: 60 },
    },
    {
      id: 'unknown', name: 'Not in the wiki', enabled: true, color: '#f0b429', label: 'NEW?',
      highlight: 'dot',
      when: { unknown: true },
    },
    {
      id: 'vendor', name: 'Vendor / essence fodder', enabled: true, color: '#6b7a73', label: 'JUNK',
      highlight: 'dot',
      when: { maxAvgRoll: 35, minTopRolls: 0 },
    },
  ];
}

/**
 * First enabled rule that matches, or null.
 *
 * A muted rule still MATCHES — the match carries `mute: true` and callers honour it. Returning null for
 * a mute would make "the filter deliberately silenced this" indistinguishable from "no rule claimed
 * it", which is exactly the distinction a filter author is debugging when a rule looks dead.
 */
export function matchLoot(item: OwnedGear, rules: readonly LootRule[], context: LootContext): LootMatch | null {
  for (const rule of rules) {
    if (!rule.enabled) continue;
    if (!matchesCondition(item, rule.when, context)) continue;
    return {
      ruleId: rule.id,
      name: rule.name,
      color: rule.color,
      label: rule.label ?? rule.name,
      highlight: rule.highlight ?? 'dot',
      sound: rule.sound ?? null,
      mute: Boolean(rule.mute),
    };
  }
  return null;
}

export function matchesCondition(item: OwnedGear, when: LootCondition, context: LootContext): boolean {
  if (when.slotTypes?.length && !when.slotTypes.includes(item.slotType)) return false;
  if (when.minTopRolls !== undefined && item.topRolls < when.minTopRolls) return false;
  if (when.maxTopRolls !== undefined && item.topRolls > when.maxTopRolls) return false;
  if (when.minRefine !== undefined && item.refine < when.minRefine) return false;
  if (when.unknown !== undefined && Boolean(item.unknown) !== when.unknown) return false;
  if (when.favorite !== undefined && item.favorite !== when.favorite) return false;

  if (when.minAvgRoll !== undefined && (item.avgRoll ?? -1) < when.minAvgRoll) return false;
  // An unplaceable average must not slip through a "junk" ceiling — treat it as unknown, not as 0.
  if (when.maxAvgRoll !== undefined && (item.avgRoll === null || item.avgRoll > when.maxAvgRoll)) return false;

  if (when.nameContains) {
    const needle = when.nameContains.toLowerCase();
    if (!item.name.toLowerCase().includes(needle)) return false;
  }

  if (when.hasChaos !== undefined) {
    const hasChaos = item.lines.some((line) => line.isChaos);
    if (hasChaos !== when.hasChaos) return false;
  }

  if (when.overRoll !== undefined) {
    const over = item.lines.some((line) => line.over);
    if (over !== when.overRoll) return false;
  }

  if (when.stats?.length) {
    const mode = when.statMode ?? 'all';
    let hits = 0;
    for (const condition of when.stats) {
      /**
       * Stat names match without regard to case.
       *
       * These are hand-typed into a rule, and the catalog's spelling is `Dex` while every player will
       * write `DEX` at least once. An exact comparison made that rule match NOTHING, with no error and
       * nothing on screen to say why — the rule simply sat there looking enabled. A filter that fails
       * silently on capitalisation is a trap, and nothing here needs `Dex` and `DEX` to be different.
       */
      const wanted = condition.stat.toLowerCase();
      const line = item.lines.find((candidate) => candidate.stat.toLowerCase() === wanted);
      const ok = Boolean(line)
        && (condition.minRollPct === undefined
          || (line!.rollPct !== null && line!.rollPct >= condition.minRollPct))
        && (condition.minValue === undefined || line!.base >= condition.minValue);
      if (ok) hits++;
      else if (mode === 'all') return false;
    }
    if (mode === 'any' && hits === 0) return false;
  }

  if (when.minSharedStats !== undefined) {
    const worn = context.wornStats;
    if (!worn) return false;
    let shared = 0;
    // Same spelling rule as above: the worn set comes from the same catalog today, but a caller that
    // assembles it differently must not silently score zero shared stats.
    const wornLower = new Set([...worn].map((stat) => stat.toLowerCase()));
    for (const line of item.lines) if (wornLower.has(line.stat.toLowerCase())) shared++;
    if (shared < when.minSharedStats) return false;
  }

  if (when.verdicts?.length) {
    const verdict = item.uid ? context.verdictByUid?.get(item.uid) : undefined;
    if (!verdict || !when.verdicts.includes(verdict)) return false;
  }

  return true;
}

/**
 * Repair a rule list from disk: drop nonsense, keep order, never throw on a hand-edited file.
 *
 * It is also the migration for rule files written before this engine was presentation-only:
 *
 *   `flash: true`         -> `highlight: 'glow'`  (the animation the flag used to mean)
 *   `action: 'dismantle'` -> `mute: true`         (see below)
 *
 * The action mapping is a judgement, stated so it can be argued with: a player who typed out
 * "dismantle this" was describing a bucket of items they do not want, and the nearest thing this engine
 * can still express about it is "stop showing it to me". The rule keeps its colour and tag, so nothing
 * is lost that cannot be turned back on by clearing one checkbox — and no configuration on disk can
 * make anything happen to an item any more.
 */
export function normalizeLootRules(input: unknown): LootRule[] {
  if (!Array.isArray(input)) return defaultLootRules();
  const rules: LootRule[] = [];
  const seen = new Set<string>();
  for (const raw of input) {
    if (!raw || typeof raw !== 'object') continue;
    const candidate = raw as Partial<LootRule> & { flash?: unknown; action?: unknown };
    const id = typeof candidate.id === 'string' && candidate.id ? candidate.id : `rule-${rules.length + 1}`;
    if (seen.has(id)) continue;
    seen.add(id);
    const highlight: LootHighlight = LOOT_HIGHLIGHTS.includes(candidate.highlight as LootHighlight)
      ? candidate.highlight as LootHighlight
      : candidate.flash ? 'glow' : 'dot';
    const sound = normalizeSoundName(candidate.sound);
    rules.push({
      id,
      name: typeof candidate.name === 'string' && candidate.name ? candidate.name.slice(0, 40) : id,
      enabled: candidate.enabled !== false,
      color: /^#[0-9a-f]{6}$/i.test(String(candidate.color)) ? String(candidate.color) : '#4ade80',
      ...(candidate.label ? { label: String(candidate.label).slice(0, 12) } : {}),
      highlight,
      ...(sound ? { sound } : {}),
      ...(candidate.mute === true || candidate.action === 'dismantle' ? { mute: true } : {}),
      when: normalizeCondition(candidate.when),
    });
  }
  return rules.length ? rules : defaultLootRules();
}

function normalizeCondition(input: unknown): LootCondition {
  const raw = (input ?? {}) as Partial<LootCondition>;
  const when: LootCondition = {};
  if (Array.isArray(raw.slotTypes)) when.slotTypes = raw.slotTypes.filter((value): value is string => typeof value === 'string');
  for (const key of ['minTopRolls', 'maxTopRolls', 'minAvgRoll', 'maxAvgRoll', 'minRefine', 'minSharedStats'] as const) {
    const value = Number(raw[key]);
    if (Number.isFinite(value)) when[key] = value;
  }
  if (typeof raw.overRoll === 'boolean') when.overRoll = raw.overRoll;
  if (typeof raw.unknown === 'boolean') when.unknown = raw.unknown;
  if (typeof raw.hasChaos === 'boolean') when.hasChaos = raw.hasChaos;
  if (typeof raw.favorite === 'boolean') when.favorite = raw.favorite;
  if (typeof raw.nameContains === 'string' && raw.nameContains) when.nameContains = raw.nameContains.slice(0, 40);
  if (raw.statMode === 'any' || raw.statMode === 'all') when.statMode = raw.statMode;
  if (Array.isArray(raw.stats)) {
    when.stats = raw.stats
      .filter((entry): entry is StatCondition => Boolean(entry) && typeof (entry as StatCondition).stat === 'string')
      .map((entry) => {
        const min = Number(entry.minRollPct);
        const value = Number(entry.minValue);
        return {
          stat: entry.stat,
          ...(Number.isFinite(min) ? { minRollPct: Math.max(0, Math.min(100, min)) } : {}),
          ...(Number.isFinite(value) ? { minValue: value } : {}),
        };
      });
  }
  if (Array.isArray(raw.verdicts)) {
    const allowed: Verdict[] = ['upgrade', 'better-rolls', 'sidegrade', 'worse'];
    when.verdicts = raw.verdicts.filter((value): value is Verdict => allowed.includes(value as Verdict));
  }
  return when;
}

/** A bag snapshot reduced to what identity comparison needs. */
export type OwnedIndex = ReadonlyMap<string, string>;

export function indexOwned(gear: readonly OwnedGear[]): Map<string, string> {
  const index = new Map<string, string>();
  for (const item of gear) if (item.uid) index.set(item.uid, item.itemId);
  return index;
}

/**
 * What changed between two bag snapshots.
 *
 * Item UIDs are stable per instance, so an added UID is genuinely a NEW item — the loot event. This is
 * why loot detection needs no extra protocol work: the server re-sends the character on every bag
 * mutation, so a diff of consecutive snapshots is the pickup, complete with its rolls.
 *
 * The FIRST snapshot is not a loot event. Treating a freshly-attached session as "you just looted 400
 * items" would bury the player in flashes, so callers pass `seeded: false` for the first one.
 */
export function diffOwned(previous: OwnedIndex | null, current: readonly OwnedGear[]): { added: OwnedGear[]; removedUids: string[] } {
  if (!previous) return { added: [], removedUids: [] };
  const added: OwnedGear[] = [];
  const seen = new Set<string>();
  for (const item of current) {
    if (!item.uid) continue;
    seen.add(item.uid);
    if (!previous.has(item.uid)) added.push(item);
  }
  const removedUids: string[] = [];
  for (const uid of previous.keys()) if (!seen.has(uid)) removedUids.push(uid);
  return { added, removedUids };
}
